### PR TITLE
Implement Grpc

### PR DESCRIPTION
**Pull request by ajax-yurkovskyi-t**

### CODE REVIEW

File: gateway/src/main/kotlin/com/example/gateway/client/NatsClient.kt [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
 import org.springframework.stereotype.Component
+import reactor.core.publisher.Flux
 import reactor.core.publisher.Mono
 
 @Component
-class NatsClient(private val natsConnection: Connection) {
-
+class NatsClient(
+    private val natsConnection: Connection,
+    private val dispatcher: Dispatcher
```
Comment (ajax-marchenko-s) [inline gateway/src/main/kotlin/com/example/gateway/client/NatsClient.kt:17]: add comma

File: gateway/src/main/kotlin/com/example/gateway/exception/GrpcExceptionHandler.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import com.google.rpc.Code
+import com.google.rpc.ErrorInfo
+import io.grpc.StatusRuntimeException
+import io.grpc.protobuf.StatusProto
+import net.devh.boot.grpc.server.advice.GrpcAdvice
+import net.devh.boot.grpc.server.advice.GrpcExceptionHandler
+import com.google.rpc.Status as RpcStatus
+
+@GrpcAdvice
+class GrpcExceptionHandler {
```
Comment (ajax-marchenko-s) [inline gateway/src/main/kotlin/com/example/gateway/exception/GrpcExceptionHandler.kt:14]: add default handler too

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/kafka/consumer/DeviceUpdateConsumerForNats.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kafka.receiver.KafkaReceiver
+import reactor.kotlin.core.publisher.toMono
+
+@Component
+class DeviceUpdateConsumerForNats(
+    private val updateDeviceForNatsKafkaReceiver: KafkaReceiver<String, ByteArray>,
+    private val natsConnection: Connection
```
Comment (ajax-marchenko-s) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/kafka/consumer/DeviceUpdateConsumerForNats.kt:17]: add comma

**Review by ajax-marchenko-s — COMMENTED** [review]
Great work!

File: grpc-api/src/main/proto/com/example/grpcapi/service/device_service.proto [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+service DeviceService {
+  rpc createDevice(grpcapi.reqrep.device.CreateDeviceRequest) returns
+      (grpcapi.reqrep.device.CreateDeviceResponse);
+
+  rpc getDeviceById(grpcapi.reqrep.device.GetDeviceByIdRequest) returns
+      (reqrep.device.GetDeviceByIdResponse);
+
+  rpc subscribeToUpdateByUserId (grpcapi.reqrep.device.GetUpdatedDevicesRequest)
+      returns (stream internal.input.reqreply.device.UpdateDeviceResponse);
```
Comment (ajax-makarytskyi-o) [inline grpc-api/src/main/proto/com/example/grpcapi/service/device_service.proto:18]: Better change response from internal.input.reqreply.device.UpdateDeviceResponse to commonmodels.device.Device.

File: gateway/src/main/kotlin/com/example/gateway/client/NatsClient.kt [hunk 5]
```diff
@@ -17,4 +24,22 @@ class NatsClient(private val natsConnection: Connection) {
         return Mono.fromFuture { natsConnection.request(subject, payload.toByteArray()) }
             .map { response -> parser.parseFrom(response.data) }
     }
+
+    fun subscribeByUserId(userId: String): Flux<UpdateDeviceResponse> {
```
Comment (ajax-shulepa-a) [inline gateway/src/main/kotlin/com/example/gateway/client/NatsClient.kt:28]: strange signature: method name defines that you use pub-sub logic, but the return type relates to req-rep logic

File: internal-api/src/main/kotlin/com/example/internal/NatsSubject.kt [hunk 6]
```diff
@@ -11,5 +11,7 @@ object NatsSubject {
         const val GET_ALL = "$DEVICE_PREFIX.get_all"
         const val UPDATE = "$DEVICE_PREFIX.update"
         const val DELETE = "$DEVICE_PREFIX.delete"
+        const val GET_BY_USER_ID = "$DEVICE_PREFIX.get_by_user_id"
+        fun updateByUserId(userId: String) = "$DEVICE_PREFIX.$userId"
```
Comment (ajax-shulepa-a) [inline internal-api/src/main/kotlin/com/example/internal/NatsSubject.kt:15]: What does this subject look like and how do we understand what it is responsible for (without looking at the function name)?

File: iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/controller/nats/device/GetDevicesByUserIdNatsController.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import com.example.iotmanagementdevice.mapper.GetDevicesByUserIdMapper
+import com.example.iotmanagementdevice.service.user.UserService
+import com.google.protobuf.Parser
+import io.nats.client.Connection
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.toMono
+
+@Component
+class GetDevicesByUserIdNatsController(
```
Comment (ajax-shulepa-a) [inline iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/controller/nats/device/GetDevicesByUserIdNatsController.kt:16]: is it covered with test?

File: gateway/src/main/kotlin/com/example/gateway/exception/GrpcExceptionHandler.kt [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    @GrpcExceptionHandler(EntityNotFoundException::class)
```
[the remark this answers is not in the document]
Comment (ajax-yurkovskyi-t) [reply gateway/src/main/kotlin/com/example/gateway/exception/GrpcExceptionHandler.kt:16]: add message to RpcStatus

File: gateway/src/main/kotlin/com/example/gateway/mapper/grpc/GetDeviceByIdGrpcMapper.kt [hunk 9]
```diff
@@ -23,15 +23,19 @@ class GetDeviceByIdGrpcMapper {
...
         val message = response.failure.message.orEmpty()
-        throw when (response.failure.errorCase!!) {
-            GetDeviceByIdResponse.Failure.ErrorCase.DEVICE_NOT_FOUND -> EntityNotFoundException(message)
-            GetDeviceByIdResponse.Failure.ErrorCase.ERROR_NOT_SET -> error(message)
-        }
+        return GrpcGetDeviceByIdResponse.newBuilder().apply {
+            failureBuilder.message = message
+            when (response.failure.errorCase!!) {
+                GetDeviceByIdResponse.Failure.ErrorCase.DEVICE_NOT_FOUND -> failureBuilder.deviceNotFoundBuilder
+                GetDeviceByIdResponse.Failure.ErrorCase.ERROR_NOT_SET -> Error.getDefaultInstance()
```
Comment (ajax-shulepa-a) [inline gateway/src/main/kotlin/com/example/gateway/mapper/grpc/GetDeviceByIdGrpcMapper.kt:37]: why do you need Error.getDefaultInstance()?

File: gateway/src/test/kotlin/com/example/gateway/mapper/grpc/GetDeviceByIdGrpcMapperTest.kt [hunk 10]
```diff
@@ -46,38 +47,48 @@ class GetDeviceByIdGrpcMapperTest {
...
+    fun `should throw RuntimeException when response not set`() {
+        val response = GetDeviceByIdResponse.getDefaultInstance()
+
         // WHEN & THEN
         val exception = assertThrows<RuntimeException> {
             getDeviceByIdGrpcMapper.toGrpc(response)
         }
 
-        assertEquals(expectedMessage, exception.message)
+        assertEquals("No response case set", exception.message)
```
Comment (ajax-shulepa-a) [inline gateway/src/test/kotlin/com/example/gateway/mapper/grpc/GetDeviceByIdGrpcMapperTest.kt:59]: we discussed that asserting exception messages is bad practice

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**